src/ucp/tag/tag_recv.c

```diff
@@ -107,42 +114,30 @@ ucp_tag_recv_request_completed(ucp_request_t *req, ucp_tag_recv_callback_t cb,
     ucs_trace_req("%s returning completed request %p (%p) stag 0x%"PRIx64" len %zu, %s",
                   function, req, req + 1, info->sender_tag, info->length,
                   ucs_status_string(status));
-    cb(req + 1, status, &req->recv.info);
 
     UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX,
                           "ucp_tag_recv_request_completed",
                           req, info->length);
-    ucp_request_put(req);
+    cb(req + 1, status, &req->recv.info);
+    ucp_request_put(req, status);
 }
 
-ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count,
-                                 uintptr_t datatype, ucp_tag_t tag, ucp_tag_t tag_mask,
-                                 ucp_tag_recv_callback_t cb)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_tag_recv_common(ucp_worker_h worker, void *buffer, size_t count,
+                    uintptr_t datatype, ucp_tag_t tag, ucp_tag_t tag_mask,
+                    ucp_request_t *req)
 {
     ucs_status_t status;
-    ucp_request_t *req;
     unsigned save_rreq = 1;
 
-    ucs_trace_req("recv_nb buffer %p count %zu tag %"PRIx64"/%"PRIx64, buffer,
-                  count, tag, tag_mask);
-
-    req = ucp_tag_recv_request_get(worker, buffer, count, datatype);
-    if (req == NULL) {
-        return UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-    }
-
-    UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX, "ucp_tag_recv_nb",
-                          req, ucp_contig_dt_length(datatype, count));
-
-    req->recv.cb = cb;
+    ucs_trace_req("recv_nb%c buffer %p count %zu tag %"PRIx64"/%"PRIx64,
+                  (req->flags & UCP_REQUEST_FLAG_EXTERNAL) ? 'r' : ' ',
+                  buffer, count, tag, tag_mask);
 
     /* First, search in unexpected list */
     status = ucp_tag_search_unexp(worker, buffer, count, datatype, tag,
                                   tag_mask, req, &req->recv.info, &save_rreq);
-    if (status != UCS_INPROGRESS) {
-        ucp_tag_recv_request_completed(req, cb, status, &req->recv.info,
-                                       "recv_nb");
-    } else if (save_rreq) {
+    if (save_rreq && status == UCS_INPROGRESS) {
```

Comment:
this change add up to 2 branches to fast path - in expected case, instead of checking once and completing the request, there are 2 more checks here. Need to check for expected case first.